ehsm_kms_service/delete_secret_thread.js

```diff
@@ -0,0 +1,137 @@
+const {
+    base64_decode
+} = require('./function')
+
+// Example of setInterval
+var init_interval
+
+// Record timer waiting to run
+var timer_queue = {}
+
+/**
+ * remove a timer from exist timer by key
+ * @param {Object} DB : database control
+ * @param {Object} key : key form exist timer
+ * @returns
+ */
+const removeTimer = (key, DB) => {
+    try {
+        //Delete thread when thread exists
+        if (timer_queue[key] != undefined ) {
+            clearTimeout(timer_queue[key].timer)
+        }
+        delete timer_queue[key];
+        // An insurance mechanism to ensure that thread init_interval is active
+        init_secret_delete_thread(DB)
+    } catch (error) {
+        console.error('error removeTimer :: ', error)
+    }
+}
+
+/**
+ * Expose the interface externally, scheduled tasks of delete secret
+ * @param {String} appid : appid of user
+ * @param {Object} DB : database control
+ * @param {Object} secretName : The name of the secret. eg. 'secretName01'
+ * @param {Object} plannedDeleteTime : planned delete time for secret eg. '1659931644416'
+ * @returns
+ */
+const startDeleteSecretTask = (DB, appid, secretName, plannedDeleteTime) => {
+    try {
+        let delayMillisecond = plannedDeleteTime - new Date().getTime()
+        let needAdd = true
+        // process exist timer, 
+        // If the new time is within plus or minus 10 seconds of the original time, keep the original plan unchanged; otherwise, clear the original plan and add again
+        if (timer_queue.hasOwnProperty(secretName)) {
+            if (timer_queue[secretName].plannedDeleteTime + 10000 > plannedDeleteTime && timer_queue[secretName].plannedDeleteTime - 10000 < plannedDeleteTime) {
+                needAdd = false
+            } else {
+                clearTimeout(timer_queue[secretName].timer)
+            }
+        }
+        //If the planned delete time exceeds 12 hours, the thread will not be started
+        if (delayMillisecond > 12 * 1000 * 60 * 60) {
+            needAdd = false
+        }
+        if (needAdd) {
+            //Force delete according to planned deletion time
+            let timer = setTimeout(_deleteSecret, delayMillisecond, DB, appid, secretName)
+            timer_queue[secretName] = {
```

Comment:
what would happen if two users have the same secret name and want to delete their secret simultaneously? the timer queue[secretName] can support this scenario?

Reply:
This is a problem. The timer queue[secretName] will be changed to secretname + '_'+ appid